preprocessing/data_science_help_functions.py

```diff
@@ -189,3 +197,178 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
+        train = train[columns_to_use]
+        test = test[columns_to_use]
+    # add identifier and combine
+    train['istrain'] = 1
+    test['istrain'] = 0
+    # max_dataframe_length
+    for df in [train, test]:
+        if len(df) > max_dataframe_length:
+            df = df.head(max_dataframe_length)
+    # add identifier and combine
+    train['istrain'] = 1
+    test['istrain'] = 0
+    df_joined = pd.concat([train, test], axis=0)
+    # convert non-numerical columns to integers
+    df_numeric = df_joined.select_dtypes(exclude=['object', 'datetime'])
+    df_obj = df_joined.select_dtypes(include=['object', 'datetime']).copy()
+    for c in df_obj:
+        df_obj[c] = pd.factorize(df_obj[c])[0]
+    df_joined = pd.concat([df_numeric, df_obj], axis=1)
+    return df_joined
+
+
+def get_adv_validation_score(df_joined: pd.DataFrame,
+                             y: pd.Series) -> Tuple[float, xgb.sklearn.XGBClassifier]:
+    """ Advisarial validation score estimator
+
+    Calculate advisarial validation score based on dataframes and XGBClassifier
+
+    :param DataFrame df_joined: Feature dataframe
+    :param Series y: Target series
+
+    :return:
+            - clf: Trained model
+            - mean of KFold validation results (ROC-AUC scores)
+    """
+
+    skf = StratifiedKFold(n_splits=3, shuffle=True, random_state=44)
+    xgb_params = {
+        'learning_rate': 0.1, 'objective': 'binary:logistic',
+        'silent': 1, 'n_estimators': 50
+    }
+    clf = xgb.XGBClassifier(**xgb_params, seed=10)
+    results = []
+    logger.info('Adversarial validation checking:')
+    for fold, (train_index, test_index) in enumerate(skf.split(df_joined, y)):
+        fold_xtrain, fold_xval = df_joined.iloc[train_index], df_joined.iloc[test_index]
+        fold_ytrain, fold_yval = y.iloc[train_index], y.iloc[test_index]
+        clf.fit(fold_xtrain, fold_ytrain, eval_set=[(fold_xval, fold_yval)],
+                eval_metric='logloss', verbose=False, early_stopping_rounds=10)
+        fold_ypred = clf.predict_proba(fold_xval)[:, 1]
+        fold_score = roc_auc_score(fold_yval, fold_ypred)
+        results.append(fold_score)
+        logger.info(f"Fold: {fold + 1} shape: {fold_xtrain.shape} score: {fold_score}")
+
+    return round(np.mean(results), 2), clf
+
+
+def xgb_important_features(xgb: xgb.sklearn.XGBClassifier,
+                           top_features: int = 5) -> str:
+    """ Important features extractor
+
+    Get top of the most important features from a trained model
+
+    :param XGBClassifier xgb: A trained model
+    :param int top_features: Max length of features to send back
+
+    :return:
+            - A string with a list of the most important features plus their importance
+    """
+
+    # get features
+    feat_imp = xgb.get_booster().get_score(importance_type='gain')
+
+    sorted_x = round_and_sort_dict(feat_imp)
+
+    return str(list(sorted_x[:top_features]))
+
+
+def round_and_sort_dict(feat_imp: dict) -> dict:
```

Comment:
Pycharm says that this function should return a dict but it got a list. Is the output a list?

Reply:
string - the purpose to print the first top_features